File: src/scan.cpp
#include "scan.hpp"

#include <algorithm>
#include <limits>
#include <new>

namespace {
struct Int2 {
    int x;
    int z;
};

Int2 rotateXzOffset(int x, int z, int direction)
{
    switch (direction / 90) {
    case 1:
        return { -z, x };
    case 2:
        return { -x, -z };
    case 3:
        return { z, -x };
    case 0:
    default:
        return { x, z };
    }
}

std::uint64_t span(int start, int end)
{
    return static_cast<std::uint64_t>(static_cast<std::int64_t>(end) - start);
}

std::uint64_t saturatedMultiply(std::uint64_t lhs, std::uint64_t rhs, bool* saturated)
{
    if (lhs != 0 && rhs > std::numeric_limits<std::uint64_t>::max() / lhs) {
        if (saturated) {
            *saturated = true;
        }
        return std::numeric_limits<std::uint64_t>::max();
    }
    return lhs * rhs;
}

std::uint64_t saturatedAdd(std::uint64_t lhs, std::uint64_t rhs, bool* saturated)
{
    if (rhs > std::numeric_limits<std::uint64_t>::max() - lhs) {
        if (saturated) {
            *saturated = true;
        }
        return std::numeric_limits<std::uint64_t>::max();
    }
    return lhs + rhs;
}
}

std::vector<RotationInfo> makeDirectionalFilter(
    const std::vector<RotationInfo>& filter,
    int direction)
{
    std::vector<RotationInfo> result;
    result.reserve(filter.size());

    const int quarterTurns = direction / 90;
    for (RotationInfo info : filter) {
        const Int2 rotated = rotateXzOffset(info.x, info.z, direction);
        info.x = static_cast<std::int8_t>(rotated.x);
        info.z = static_cast<std::int8_t>(rotated.z);
        if (info.visibleMask == 3) {
            info.rotation = static_cast<std::uint8_t>((info.rotation + quarterTurns) % 4);
        }
        result.push_back(info);
    }
    // Four-state samples reject three quarters of candidates, ahead of two-state side samples.
    // Preserve the config order within each group so equivalent filters remain deterministic.
    std::stable_sort(result.begin(), result.end(), [](const RotationInfo& lhs, const RotationInfo& rhs) {
        return lhs.visibleMask > rhs.visibleMask;
    });
    return result;
}

std::uint64_t workItemCandidateCount(const WorkItem& item, bool* saturated)
{
    bool overflow = false;
    const std::uint64_t xy = saturatedMultiply(
        span(item.start.x, item.end.x),
        span(item.start.y, item.end.y),
        &overflow);
    const std::uint64_t xyz = saturatedMultiply(xy, span(item.start.z, item.end.z), &overflow);
    if (saturated) {
        *saturated = overflow;
    }
    return xyz;
}

bool makeScanPlan(
    const ScanConfig& config,
    TileSize tileSize,
    ScanPlan* plan,
    std::string* error)
{
    if (!plan) {
        if (error) {
            *error = "missing scan plan output";
        }
        return false;
    }

    if (tileSize.x <= 0 || tileSize.z <= 0) {
        if (error) {
            *error = "tile dimensions must be positive";
        }
        return false;
    }

    const std::uint64_t xSpan = span(config.xRange.start, config.xRange.end);
    const std::uint64_t zSpan = span(config.zRange.start, config.zRange.end);
    const std::uint64_t tileX = static_cast<std::uint64_t>(tileSize.x);
    const std::uint64_t tileZ = static_cast<std::uint64_t>(tileSize.z);
    const std::uint64_t xTiles = 1 + (xSpan - 1) / tileX;
    const std::uint64_t zTiles = 1 + (zSpan - 1) / tileZ;

    bool workCountOverflow = false;
    const std::uint64_t tileCount = saturatedMultiply(xTiles, zTiles, &workCountOverflow);
    const std::uint64_t workCount = saturatedMultiply(
        tileCount,
        static_cast<std::uint64_t>(config.directions.size()),
        &workCountOverflow);

    if (workCountOverflow || workCount > plan->items.max_size()) {
        if (error) {
            *error = "scan contains too many work items for this build";
        }
        return false;
    }

    ScanPlan built;
    try {
        built.items.reserve(static_cast<std::size_t>(workCount));
    }
    catch (const std::bad_alloc&) {
        if (error) {
            *error = "not enough memory to build the scan plan";
        }
        return false;
    }

    // Tiles are anchored at the minimum bounds; edge tiles retain exclusive ends.
    auto addTile = [&](std::uint64_t tileIndexX, std::uint64_t tileIndexZ, std::size_t directionIndex) {
        const std::int64_t xStart = static_cast<std::int64_t>(config.xRange.start)
            + static_cast<std::int64_t>(tileIndexX * tileX);
        const std::int64_t zStart = static_cast<std::int64_t>(config.zRange.start)
            + static_cast<std::int64_t>(tileIndexZ * tileZ);
        const std::int64_t xEnd = std::min<std::int64_t>(
            xStart + tileSize.x,
            config.xRange.end);
        const std::int64_t zEnd = std::min<std::int64_t>(
            zStart + tileSize.z,
            config.zRange.end);

        WorkItem item = {
            { static_cast<int>(xStart), config.yRange.start, static_cast<int>(zStart) },
            { static_cast<int>(xEnd), config.yRange.end, static_cast<int>(zEnd) },
            directionIndex,
            config.directions[directionIndex],
        };
        bool itemSaturated = false;
        const std::uint64_t candidates = workItemCandidateCount(item, &itemSaturated);
        built.totalCandidates = saturatedAdd(
            built.totalCandidates,
            candidates,
            &built.totalCandidatesSaturated);
        built.totalCandidatesSaturated = built.totalCandidatesSaturated || itemSaturated;
        built.items.push_back(item);
    };

    if (config.scanOrder == ScanOrder::Linear) {
        // X -> Z -> Direction ordering.
        for (std::uint64_t x = 0; x < xTiles; ++x) {
            for (std::uint64_t z = 0; z < zTiles; ++z) {
                for (std::size_t direction = 0; direction < config.directions.size(); ++direction) {
                    addTile(x, z, direction);
                }
            }
        }
    }
    else {
        const std::int64_t centerX = static_cast<std::int64_t>((xSpan - 1) / 2 / tileX);
        const std::int64_t centerZ = static_cast<std::int64_t>((zSpan - 1) / 2 / tileZ);
        const std::int64_t maxRadius = std::max({
            centerX,
            centerZ,
            static_cast<std::int64_t>(xTiles - 1) - centerX,
            static_cast<std::int64_t>(zTiles - 1) - centerZ,
        });

        // Spiral mode is tile-major: check every direction before moving farther out.
        auto emit = [&](std::int64_t x, std::int64_t z) {
            if (x < 0 || z < 0 || x >= static_cast<std::int64_t>(xTiles) || z >= static_cast<std::int64_t>(zTiles)) {
                return;
            }
            for (std::size_t direction = 0; direction < config.directions.size(); ++direction) {
                addTile(static_cast<std::uint64_t>(x), static_cast<std::uint64_t>(z), direction);
            }
        };

        emit(centerX, centerZ);
        for (std::int64_t radius = 1; radius <= maxRadius; ++radius) {
            // Walk clockwise around each ring. Each side is clipped before iteration,
            // avoiding quadratic work for very long, narrow search rectangles.
            const std::int64_t right = centerX + radius;
            if (right >= 0 && right < static_cast<std::int64_t>(xTiles)) {
                const std::int64_t from = std::max<std::int64_t>(0, centerZ - radius + 1);
                const std::int64_t to = std::min<std::int64_t>(static_cast<std::int64_t>(zTiles) - 1, centerZ + radius);
                for (std::int64_t z = from; z <= to; ++z) {
                    emit(right, z);
                }
            }

            const std::int64_t bottom = centerZ + radius;
            if (bottom >= 0 && bottom < static_cast<std::int64_t>(zTiles)) {
                const std::int64_t from = std::min<std::int64_t>(static_cast<std::int64_t>(xTiles) - 1, centerX + radius - 1);
                const std::int64_t to = std::max<std::int64_t>(0, centerX - radius);
                for (std::int64_t x = from; x >= to; --x) {
                    emit(x, bottom);
                }
            }

            const std::int64_t left = centerX - radius;
            if (left >= 0 && left < static_cast<std::int64_t>(xTiles)) {
                const std::int64_t from = std::min<std::int64_t>(static_cast<std::int64_t>(zTiles) - 1, centerZ + radius - 1);
                const std::int64_t to = std::max<std::int64_t>(0, centerZ - radius);
                for (std::int64_t z = from; z >= to; --z) {
                    emit(left, z);
                }
            }

            const std::int64_t top = centerZ - radius;
            if (top >= 0 && top < static_cast<std::int64_t>(zTiles)) {
                const std::int64_t from = std::max<std::int64_t>(0, centerX - radius + 1);
                const std::int64_t to = std::min<std::int64_t>(static_cast<std::int64_t>(xTiles) - 1, centerX + radius);
                for (std::int64_t x = from; x <= to; ++x) {
                    emit(x, top);
                }
            }
        }
    }

    if (built.items.size() != static_cast<std::size_t>(workCount)) {
        if (error) {
            *error = "internal error: scan order did not cover every tile exactly once";
        }
        return false;
    }

    *plan = std::move(built);
    return true;
}

File: src/scan.rs
use crate::config::{ScanConfig, ScanOrder, TileSize, rotate_xz};
use crate::types::{Int3, RotationInfo};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct WorkItem {
    pub start: Int3,
    pub end: Int3,
    pub direction_index: usize,
    pub direction: i32,
}

#[derive(Clone, Debug, Default)]
pub struct ScanPlan {
    pub items: Vec<WorkItem>,
    pub total_candidates: u64,
    pub total_candidates_saturated: bool,
}

pub fn directional_filter(filter: &[RotationInfo], direction: i32) -> Vec<RotationInfo> {
    let mut result: Vec<_> = filter
        .iter()
        .copied()
        .map(|mut item| {
            let (x, z) = rotate_xz(i32::from(item.x), i32::from(item.z), direction);
            item.x = x as i8;
            item.z = z as i8;
            if item.visible_mask == 3 {
                item.rotation = (item.rotation + (direction / 90) as u8) % 4;
            }
            item
        })
        .collect();
    result.sort_by_key(|item| std::cmp::Reverse(item.visible_mask));
    result
}

fn span(start: i32, end: i32) -> u64 {
    (i64::from(end) - i64::from(start)) as u64
}

pub fn candidate_count(item: &WorkItem) -> (u64, bool) {
    let mut count = 1_u64;
    for dimension in [
        span(item.start.x, item.end.x),
        span(item.start.y, item.end.y),
        span(item.start.z, item.end.z),
    ] {
        match count.checked_mul(dimension) {
            Some(value) => count = value,
            None => return (u64::MAX, true),
        }
    }
    (count, false)
}

pub fn make_plan(config: &ScanConfig, tile_size: TileSize) -> Result<ScanPlan, String> {
    if tile_size.x <= 0 || tile_size.z <= 0 {
        return Err("tile dimensions must be positive".to_owned());
    }
    let x_span = span(config.x_range.start, config.x_range.end);
    let z_span = span(config.z_range.start, config.z_range.end);
    let tile_x = tile_size.x as u64;
    let tile_z = tile_size.z as u64;
    let x_tiles = x_span.div_ceil(tile_x);
    let z_tiles = z_span.div_ceil(tile_z);
    let work_count = x_tiles
        .checked_mul(z_tiles)
        .and_then(|count| count.checked_mul(config.directions.len() as u64))
        .ok_or_else(|| "scan contains too many work items".to_owned())?;
    let capacity = usize::try_from(work_count)
        .map_err(|_| "scan contains too many work items for this build".to_owned())?;
    let mut plan = ScanPlan {
        items: Vec::with_capacity(capacity),
        ..ScanPlan::default()
    };

    let mut add_tile = |tile_index_x: u64, tile_index_z: u64, direction_index: usize| {
        let x_start = i64::from(config.x_range.start) + (tile_index_x * tile_x) as i64;
        let z_start = i64::from(config.z_range.start) + (tile_index_z * tile_z) as i64;
        let item = WorkItem {
            start: Int3 {
                x: x_start as i32,
                y: config.y_range.start,
                z: z_start as i32,
            },
            end: Int3 {
                x: (x_start + tile_x as i64).min(i64::from(config.x_range.end)) as i32,
                y: config.y_range.end,
                z: (z_start + tile_z as i64).min(i64::from(config.z_range.end)) as i32,
            },
            direction_index,
            direction: config.directions[direction_index],
        };
        let (candidates, saturated) = candidate_count(&item);
        match plan.total_candidates.checked_add(candidates) {
            Some(total) if !saturated => plan.total_candidates = total,
            _ => {
                plan.total_candidates = u64::MAX;
                plan.total_candidates_saturated = true;
            }
        }
        plan.items.push(item);
    };

    let mut emit = |x: i64, z: i64| {
        if x >= 0 && z >= 0 && x < x_tiles as i64 && z < z_tiles as i64 {
            for direction in 0..config.directions.len() {
                add_tile(x as u64, z as u64, direction);
            }
        }
    };

    match config.scan_order {
        ScanOrder::Linear => {
            for x in 0..x_tiles {
                for z in 0..z_tiles {
                    emit(x as i64, z as i64);
                }
            }
        }
        ScanOrder::Spiral => {
            let center_x = ((x_span - 1) / 2 / tile_x) as i64;
            let center_z = ((z_span - 1) / 2 / tile_z) as i64;
            let max_radius = [
                center_x,
                center_z,
                x_tiles as i64 - 1 - center_x,
                z_tiles as i64 - 1 - center_z,
            ]
            .into_iter()
            .max()
            .unwrap_or(0);
            emit(center_x, center_z);
            for radius in 1..=max_radius {
                for z in center_z - radius + 1..=center_z + radius {
                    emit(center_x + radius, z);
                }
                for x in (center_x - radius..=center_x + radius - 1).rev() {
                    emit(x, center_z + radius);
                }
                for z in (center_z - radius..=center_z + radius - 1).rev() {
                    emit(center_x - radius, z);
                }
                for x in center_x - radius + 1..=center_x + radius {
                    emit(x, center_z - radius);
                }
            }
        }
    }
    if plan.items.len() != capacity {
        return Err("internal error: scan order did not cover every tile exactly once".to_owned());
    }
    Ok(plan)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::{IntRange, ScanConfig};
    use crate::types::TextureAlgorithm;
    use std::collections::HashSet;

    fn config() -> ScanConfig {
        ScanConfig {
            algorithm: TextureAlgorithm::Vanilla3,
            directions: vec![0, 90],
            x_range: IntRange { start: -2, end: 3 },
            y_range: IntRange { start: 0, end: 1 },
            z_range: IntRange { start: -2, end: 3 },
            cpu_tile_size: TileSize { x: 1, z: 1 },
            gpu_tile_size: TileSize { x: 1, z: 1 },
            filter: vec![RotationInfo::new(0, 0, 0, 0, false)],
            ..ScanConfig::default()
        }
    }

    #[test]
    fn rotates_and_orders_filters() {
        let filter = [
            RotationInfo::new(2, 3, -4, 1, false),
            RotationInfo::new(-1, 0, 5, 1, true),
        ];
        let rotated = directional_filter(&filter, 90);
        assert_eq!((rotated[0].x, rotated[0].z, rotated[0].rotation), (4, 2, 2));
        assert_eq!(
            (rotated[1].x, rotated[1].z, rotated[1].rotation),
            (-5, -1, 1)
        );
    }

    #[test]
    fn builds_linear_and_spiral_plans() {
        let mut config = config();
        config.scan_order = ScanOrder::Spiral;
        let spiral = make_plan(&config, config.gpu_tile_size).unwrap();
        assert_eq!(spiral.items.len(), 50);
        assert_eq!((spiral.items[0].start.x, spiral.items[0].start.z), (0, 0));
        let visited: HashSet<_> = spiral
            .items
            .iter()
            .map(|item| (item.start.x, item.start.z, item.direction))
            .collect();
        assert_eq!(visited.len(), spiral.items.len());

        config.scan_order = ScanOrder::Linear;
        let linear = make_plan(&config, config.gpu_tile_size).unwrap();
        assert_eq!((linear.items[0].start.x, linear.items[0].start.z), (-2, -2));
    }

    #[test]
    fn spiral_covers_narrow_rectangles() {
        let mut config = config();
        config.directions = vec![0];
        config.scan_order = ScanOrder::Spiral;
        config.x_range = IntRange { start: 0, end: 10 };
        config.z_range = IntRange { start: -1, end: 1 };
        let plan = make_plan(&config, TileSize { x: 3, z: 1 }).unwrap();
        let visited: HashSet<_> = plan
            .items
            .iter()
            .map(|item| (item.start.x, item.start.z))
            .collect();
        assert_eq!(plan.items.len(), 8);
        assert_eq!(visited.len(), 8);
    }
}
